### PR TITLE
Add a CanDeploy check to Transform at FrameEndTask

### LINKED ISSUE #13984: MCV Deployment overlaps with adjacent MCVs
Discovered by the player egg.

(Release & Playtest)
So long as the MCVs are all facing the same direction and at least 1 cell apart, they can deploy on top of one another if deployed at the same time:
mcv_deploy

### PR DESCRIPTION
**Pull request by PunkPun**
Closes #13984

### CODE REVIEW

**Review by abcdefg30 — DISMISSED** [review]

**Comment by anvilvapre** [comment]
won't it still show the undeploy animation? or is that no issue? it will just not deploy at the end?

**Comment by PunkPun** [comment]
The animation won't play. Try it

**Comment by PunkPun** [comment]
Oh I just realised. MCV doesn't have a deploy animation so actually I'm unsure. The animation might play but the unit won't transform.

We can't do much about this, it's a choice. Do we want the transform always after the animation stops playing or only if it's viable? 

I see 3 options
- Only transform when viable
- If there's an animation - don't check, if there's no animation - check
- Expose this to yaml and let the modder decide on what should happen upon the animation completing

Skipping the check would be as simple as passing a null into the DoTransforms() so on a technical level each option is a clean solution. Though customisation might be a bit messy considering so many traits use this activity. FWIW I'm fine with the current state of the PR

**Comment by PunkPun** [comment]
I made it so when after playing the make animation deploying is invalid, it plays it in reverse

The testcase is a bit misleading because this is the Construction Yards make animation and not the MCV's, so it's reversed. But it's enough for testing

**Comment by PunkPun** [comment]
I made it so when after playing the make animation deploying is invalid, it plays it in reverse

The testcase is a bit misleading because this is the Construction Yard's make animation and not the MCV's, so it's reversed. But it's enough for testing

https://user-images.githubusercontent.com/37534529/206706056-b21a3b42-940b-42a2-96e9-16647503bfa7.mov

**Comment by abcdefg30** [comment]
Confirmed that this still works. ✅ 

Btw, you can define
```
	make: factmake
		Length: *
		Frames: 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0
```
to fix the animation for the testcase.

File: OpenRA.Mods.Common/Activities/Transform.cs [hunk 1]
```diff
@@ -85,6 +85,15 @@ void DoTransform(Actor self)
 				if (self.IsDead || self.WillDispose)
 					return;
 
+				// Prevent deployment in bogus locations
+				if (transforms != null && !transforms.CanDeploy())
+				{
+					if (!SkipMakeAnims && makeAnimation != null)
+						makeAnimation.Forward(self, () => { IsInterruptible = true; Cancel(self, true); });
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Activities/Transform.cs:92]:
```suggestion
					if (!SkipMakeAnims && makeAnimation != null)
						makeAnimation.Forward(self, () => { IsInterruptible = true; Cancel(self, true); });
					else
						Cancel(self, true);
```
?

Comment (PunkPun) [reply OpenRA.Mods.Common/Activities/Transform.cs:92]: These are the same checks that exist in Tick so there'd be nothing to cancel when they are false. Though I guess I can add it for completeness

**Review by Mailaender — DISMISSED** [review]

**Comment by Mailaender** [comment]
Can you remove the test case? https://github.com/OpenRA/OpenRA/pull/20527/commits/62321f18e5ea969ebcbdadf3e109bf62f518f6b0

**Comment by PunkPun** [comment]
dropped the testcase

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8a339e0397e87ee02cbcc939a10a530ebcb8b4ca...a246b6452c45ff40cedcee1a8c7f68280d184753)

**Pull request merged**